source/core_sntp_client.c

```diff
@@ -0,0 +1,104 @@
+/*
+ * coreSNTP v1.0.0
+ * Copyright (C) 2021 Amazon.com, Inc. or its affiliates.  All Rights Reserved.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of
+ * this software and associated documentation files (the "Software"), to deal in
+ * the Software without restriction, including without limitation the rights to
+ * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
+ * the Software, and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+/**
+ * @file core_sntp_client.c
+ * @brief Implementation of the client API of the coreSNTP library.
+ */
+
+/* Standard includes. */
+#include <assert.h>
+#include <string.h>
+
+/* SNTP client library API include. */
+#include "core_sntp_client.h"
+
+
+SntpStatus_t Sntp_Init( SntpContext_t * pContext,
+                        const SntpServerInfo_t * pTimeServers,
+                        size_t numOfServers,
+                        uint8_t * pNetworkBuffer,
+                        size_t bufferSize,
+                        SntpResolveDns_t resolveDnsFunc,
+                        SntpGetTime_t getSystemTimeFunc,
+                        SntpSetTime_t setSystemTimeFunc,
+                        const UdpTransportIntf_t * pTransportIntf,
+                        const SntpAuthenticationIntf_t * pAuthIntf )
+{
+    SntpStatus_t status = SntpSuccess;
+
+    /* Validate pointer parameters are not NULL. */
+    if( ( pContext == NULL ) || ( pTimeServers == NULL ) || ( numOfServers == 0U ) ||
+        ( pNetworkBuffer == NULL ) || ( resolveDnsFunc == NULL ) || ( getSystemTimeFunc == NULL ) ||
+        ( setSystemTimeFunc == NULL ) || ( pTransportIntf == NULL ) )
+    {
+        status = SntpErrorBadParameter;
+    }
+    /* Validate that the members of the UDP transport interface. */
+    else if( ( pTransportIntf->recvFrom == NULL ) || ( pTransportIntf->sendTo == NULL ) )
```

Comment:
Is it necessary for `NULL` checks to be in separate `else if` statements? Once the preceding predicates between each `||` are true, the remaining predicates will no longer be evaluated. Although, this might be done for complexity or MISRA. Just curious.

Reply:
The reason for separating out the NULL checks for members of the interface objects is only for better readability, and potentially, for better logging messages that I will add in a separate PR